vignettes/slowly-changing-dimension.Rmd

```diff
@@ -0,0 +1,241 @@
+
+forecasts2 <- filter(forecasts_full, ForecastDate == "2023-09-29") %>%
+  select(!"ForecastDate")
+forecasts2a <- filter(forecasts_full, ForecastDate == "2023-09-29")
+```
+
+A slowly changing dimension is a concept in data warehousing which refers to data which may change over time, but at an irregular schedule.
+
+## Type 1 and Type 2 history
+
+For example, consider the following table of forecasts for a number of cities:
+```{r forecasts}
+# Current date: 2023-09-28
+forecasts
+```
+
+The following day, the forecasts will have changed, and — barring the occasional data hoarder — the existing data is no longer relevant.
+
+In this example, most (if not all) of the values of the `Forecast_C` column will change with each regular update.
+Putting it into other words, the table is a *snapshot*^[A *snapshot* is a static view of (part of) a database at a specific point in time] of forecasts at the last time of update.
+
+The following day, the forecasts naturally change:
+```{r forecasts2a}
+# Current date: 2023-09-29
+forecasts2
+```
+We could choose to update the forecasts table so that it would always contain the current data.
+This is what is referred to as Type 1 methodology [@Kimball2013].
+
+Databases are thankfully a rather efficient way of storing and accessing data, so instead of discarding the values from the previous day, we append the new data to those of the previous day.
+Also, in order to keep our data organised, we add a column with the date of the forecast, aptly named `ForecastDate`.
+
+The full table of forecasts for the two days now looks like below, and we are slowly building a full history of forecasts:
+```{r forecasts_full}
+forecasts_full
+```
+Managing historical data by inserting new data in this manner is often referred to as Type 2 methodology or Type 2 history.
+
+Our table now provides much more information for the user through filtering:
+```{r forecasts_full_examples}
+library(tidyverse)
+
+# Current forecasts
+forecasts_full %>%
+  slice_max(ForecastDate, n = 1) %>%
+  select(!"ForecastDate")
+
+# Forecasts for a given date
+forecasts_full %>%
+  filter(ForecastDate == "2023-09-28")
+
+# Full history for a given city
+forecasts_full %>%
+  filter(City == "New York")
+```
+
+Now, we note that the forecast for Houston has not changed between the two days.
+
+In order to keep our data as minimized as possible, we modify the table again, now expanding `ForecastDate` into `ForecastFrom` and `ForecastUntil`.
+
+Our table of forecasts now looks like this:
+```{r forecasts_scd}
+forecasts_scd
+```
+For now, the `ForecastUntil` value is set to `NA`, as it is not known when these rows will "expire" (if ever).
+This also makes it easy to identify currently valid data.
+
+Adding a new column to save a single row of data naturally seems a bit overkill, but as the number of rows in the data set increases indefinitely, this solutions scales much better.
+
+## A "timeline of timelines"
```

Comment:
You can maybe throw in the phrase "bitemporal" in this section.
https://en.wikipedia.org/wiki/Bitemporal_modeling

I have often come across this phrasing while looking up Type-2 history, so I think it will be a good phrase to include.
(for people who are familiar with the term and for SEO)

Reply:
Good idea; I've added a line mentioning the term "bitemporal modeling". Good call on the SEO! 😄